db: cql3: add comments regarding unsafe interval<clustering_key_prefix>

andrzej-jackowski-scylladb opened the pull request:
class clustering_range is a range of Clustering Key Prefixes implemented as interval<clustering_key_prefix>. However, due to the nature of Clustering Key Prefix, the ordering of clustering_range is complex and does not satisfy the invariant of interval<>. To be more specific, as a comment in interval<> implementation states: “The end bound can never be smaller than the start bound”. As a range of CKP violates the invariant, some algorithms, like intersection(), can return incorrect results. For more details refer to scylladb#8157, scylladb#21604, scylladb#22817.

This commit:
 - Add a WARNING comment to discourage usage of clustering_range
 - Add WARNING comments to potentially incorrect uses of interval<clustering_key_prefix> non-trivial methods
 - Add a FIXME comment to incorrect use of interval<clustering_key_prefix_view>::deoverlap and WARNING comments to related interval<clustering_key_prefix_view> misuse.

Only comments are changed - no need to backport

ptrsmrn reviewed (approved):
Looks good but I'd double check check with @tgrabiec. Tomek, anything else to add here?

tgrabiec commented on query-request.hh:
```diff
@@ -48,6 +48,11 @@ using range = wrapping_interval<T>;
+// Due to the note below, methods such as intersection() and deoverlap() returns unexpected (i.e. incorrect) results.
+// If only possible, do not use interval<clustering_key_prefix> in new code. Also refer to
+// optional<clustering_range> intersection(const clustering_range&, const clustering_range&, const prefix_equal_tri_compare&)
+// in statement_restrictions.cc for correct implementation of intersection() method.
```
Should suggest an alternative: position_range.

The pointer to intersection() in statement_restrictions.cc is useful, but we don't want to encourage it either, it's too complicated.

andrzej-jackowski-scylladb replied:
Added information that position_range should be used and that intersection() is overcomplicated.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15436/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6670/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3548/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14689/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15436/)
#### Failed Tests (1/37968):
- [test_random_failures[stop_after_streaming-restart_coordinator_node].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15436/testReport/junit/topology_random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_streaming_restart_coordinator_node__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_streaming-restart_coordinator_node].debug.2)
#### Build Details:
- Duration: 8 hr 9 min
- Builder: i-0c9f9a473ffd991cf (m5ad.12xlarge)

tchaikov commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15436/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6670/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3548/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14689/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15436/)
> #### Failed Tests (1/37968):
> 
>     * [test_random_failures[stop_after_streaming-restart_coordinator_node].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15436/testReport/junit/topology_random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_streaming_restart_coordinator_node__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_after_streaming-restart_coordinator_node%5D.debug.2)
> 

#21872

andrzej-jackowski-scylladb commented:
v2:
 - Modified a comment to encourage usage of position_range instead of interval<>
 - Modified a comment include information that current intersection implementation is overcomplicated.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15463/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3569/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14710/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6691/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15463/)
#### Build Details:
- Duration: 8 hr 10 min
- Builder: i-0017531156b83fbe0 (m5ad.12xlarge)

ptrsmrn commented:
@scylladb/scylla-maint please consider merging this PR.

piodul commented:
Queued
